### ISSUE TITLE
Potential considerations for using Tempel with abbrevs

### ISSUE THREAD

**Issue by alecStewart1** [minad/tempel#112]
Hello, 

This isn't so much an issue so much as it is something that I've noticed when implementing something myself that's tangentially similar to Tempel.

There's a few funky things with abbrevs in Emacs that I don't know if it's an issue in Tempel when using it with abbrevs, but I thought I'd mention them here:

1. Abbrevs will expand when a `word` is matched, _period_
 
So in [tempel-collection](https://github.com/Crandel/tempel-collection) for the Emacs Lisp templates there's `let*`, `iflet*`, and `whenlet*`,  *but* there's `let`, `iflet`, `whenlet`. The abbrev will expand with `let` and not `let*` and so on, because the asterisk counts as it's own character, it's not apart of the `word`. So if you were to do `M-f` on `let*` in Emacs, the cursor won't go to the position after the asterisk but at `let`, so it'll be `let[ here's the cursor ]*`. I'd have to see, but maybe this wouldn't be an issue in the case of prefixing a `word` you'd like to use for an abbrev/template with a character like `@`, `#` or `*`.

I don't know if this is an issue with Tempel, but I do see in the README that the template for `let*` is `lett`. People here would now better than I, but if you're using Tempel + abbrevs then I don't think having both `let` and `lett` as templates would work, as abbrevs will get so eagerly expanded. I could be wrong and this could be a total nonissue.

2. Abbrevs will expand practically everywhere.

In regular text, comments, comment blocks, strings, docstrings, you name it. The way I dealt with this was make use of `define-abbrev`'s `:expand-function` keyword argument and using this (admittedly hacky) function:

```emacs-lisp
(defun inside-comment-or-string-p ()
  "Check if the point in inside a comment or string"
  (let* ((pos (point))
         (faces (get-text-property pos 'face)))
    (when (not (listp faces))
      (setf faces (list faces)))
    (save-excursion
      (if (derived-mode-p 'org-mode)
          (save-match-data
            (beginning-of-line)
            (looking-at "^[ \t]*#"))
        (or
         (nth 4 (syntax-ppss))
         (nth 8 (syntax-ppss))
         (numberp (cl-position 'font-lock-comment-face faces))
         (numberp (cl-position 'font-lock-comment-delimiter-face faces))
         (numberp (cl-position 'font-lock-doc-face faces))
         (numberp (cl-position 'font-lock-string-face faces)))))))
```

So you could have `:expand-function (lambda (&rest _) (not (inside-comment-or-string-p)))` on an abbrev and, from what I can tell (I haven't tested every programming mode package in existence), the abbrev won't expand inside a comment, comment block, string or docstring. Unfortunately it kind of hinges on that any and every mode using `font-lock-comment-face` and the other font-lock faces if the use of `syntax-ppss` fails for stuff like comments. 

Of course, one could just disable automatically expanding abbrevs, but that sort of feels like it defeats the purpose of using them (autotyping and thus less time hitting keys).

Again, I don't know if this is an issue for using Temepl + abbrevs or not, but it's just something I've found in my own experimenting with abbrevs.

Just some things I've found. They could be useful or not.

**Comment by minad** [comment]
Hi Alec,  

thanks for your thoughts!

1. To be precise, abreviations are only auto expanded after SPC and even this requires enabling `abbrev-mode`. You can also manually invoke `expand-abbrev` via `C-x a e`. But as you say, maybe this defeats the purpose since one has to type too much. I am not sure about the auto expansion mode you are using, but if one expands immediately after `prefix` it will never work well since one cannot expand the templates named `prefixsomething`.

2. True. It would be nice to have context sensitive templates. In Tempel on can use `:when` conditions in the template group header. Unfortunately it seems that my `tempel-abbrev-mode` doesn't check these conditions all the time, only once when setting up the abbreviations. This is actually a bug in `tempel-abbrev-mode`.

It seems with Tempel most people prefer to use the Capf mechanism, including myself. There one simply has more flexibility in comparison to using the abbrev tables. Tempel offers two commands/Capfs for expansion `tempel-expand` and `tempel-complete`. The more restricted `tempel-expand` only triggers if you type the exact template name and press TAB. Furthermore Tempel has the option `tempel-trigger-prefix` which ensures that template expansion takes only place if you type the prefix and the template name.

(I close this issue since it is not really a Tempel issue, but you are of course welcome to continue the discussion!)

**Comment by alecStewart1** [comment]
> (I close this issue since it is not really a Tempel issue, but you are of course welcome to continue the discussion!)

Totally understand.

> It seems with Tempel most people prefer to use the Capf mechanism, including myself. There one simply has more flexibility in comparison to using the abbrev tables.

I guess in this case, would you recommend to tempel users to advise whatever function completes abbrevs at point? So in the case of your other package `cape`, should `cape-wrap-predicate` be made use of for the `cape-abbrev` Capf? 

So while you mentioned `tempel-expand` does specifically match a template (this is at [line 677](https://github.com/minad/tempel/blob/9656b27ff7f0144728831110514c016050fffb21/tempel.el#L677C17-L677C47) of the current main branch), one could do:

```emacs-lisp
;; this function is totally redudant but you get the idea
(defun match-tempel-template (sym)
  "Be paranoid an make doubly sure that SYM is in `tempel--templates'."
  (assq sym tempel--templates))

(setq-local completion-at-point-functions
            (list (cape-capf-predicate #'cape-abbrev
                                       #'match-tempel-template)))
```

Redudant, but it does achieve the exact match. I guess technically you would have to check if the abbrev is in the abbrev table:

```emacs-lisp
;; I know Tempel puts abbrevs in the `local-abbrev-table'
(defun abbrev-in-local-table-p (abbrev)
  "Check if ABBREV is in `local-abbrev-table'.
ABBREV must be a symbol."
  ;; likely don't need `not' or `null' but better to be specific I guess
  (not (null (abbrev-symbol (symbol-name abbrev) 'local-abbrev-table))))

(setq-local completion-at-point-functions
            (list (cape-capf-predicate #'cape-abbrev
                                       #'abbrev-in-local-table-p)))
```

**Comment by minad** [comment]
> I guess in this case, would you recommend to tempel users to advise whatever function completes abbrevs at point? So in the case of your other package cape, should cape-wrap-predicate be made use of for the cape-abbrev Capf?

I don't think I understand the question. `cape-wrap-predicate` can be used to filter the candidates with a predicate, so it may achieve what you want, but it is probably easier to just write your own dedicated Capf like `tempel-expand`.

> There's is this suggestion on the EmacsWiki to have the ability to only expand abbrevs after 2 spaces. Seems a bit silly.

I think it is just easier to trigger special expansion manually with a key. I use manual expansion for special completions commands as documented in https://github.com/minad/cape#configuration. I've also bound `tempel-complete` to an extra key. For auto completion (and `completion-at-point`) I've only set up the following Capfs globally: `tempel-expand`, `cape-elisp-block`, `cape-file`, `cape-elisp-block`, `cape-dabbrev`, `cape-dict`. These Capfs can run in addition to the standard Capf provided by the major mode. I don't use the abbrev mechanism much beside that.

**Comment by alecStewart1** [comment]
> I don't think I understand the question. cape-wrap-predicate can be used to filter the candidates with a predicate, so it may achieve what you want, but it is probably easier to just write your own dedicated Capf like tempel-expand.

Oh that's my bad. I was wondering how you'd recommend people work around:

> Unfortunately it seems that my tempel-abbrev-mode doesn't check these conditions all the time, only once when setting up the abbreviations. This is actually a bug in tempel-abbrev-mode.

^^ that, while also making sure that when typing `lett` it would specifically expand to `let*` when using `tempel-abbrev-mode`. 

>  think it is just easier to trigger special expansion manually with a key.

I might actually do that. I would need to set `abbrev-mode` to `nil` to prevent automatic expansion. Funnily enough, there's an issue I'm having with `cape-abbrev` that I'll have to put on `cape`.

**Comment by minad** [comment]
> Oh that's my bad. I was wondering how you'd recommend people work around making sure lett/let* does not expand to let template when using tempel-abbrev-mode. That seems like it would only be an issue if you've set abbrev-mode to t though.

I recommend to avoid special letters in template names, lett is better than let*. This is a problem in both cases, when you use Capf or Abbrev. Abbrev because of the auto expansion and Capf because special letters may have special meaning during completion.

Overall not many people seem to use Tempel with Abbrev, since I've never got any feedback regarding it. I also just found another bug - `no-self-insert` didn't work (anymore?). See https://github.com/minad/tempel/commit/93deb89ac1b1b6d789779432cfae5c76d9939c44. 

I think this discussion shows that Abbrev is somewhat hard to tame. Either one uses manual triggers, but then one could just as well use something different like `completion-at-point` and avoid the whole Abbrev mechanism, or one uses auto abbrev and suffers from undesired expansions.

**Comment by alecStewart1** [comment]
> I think this discussion shows that Abbrev is somewhat hard to tame. Either one uses manual triggers, but then one could just as well use something different like completion-at-point and avoid the whole Abbrev mechanism, or one uses auto abbrev and suffers from undesired expansions.

Agreed. I'll make a separate branch in my config and see if I can tinker with Tempel to get the expansion via abbrevs more nicely, if they aren't already. I will say that using `completion-at-point`, at one point, was working fine for abbrevs for me. I don't know if I made some change that's now causing me issue. You'd select the abbrev in the completion menu, and then further typing (pressing `SPC`) would expand the template. Simple! 

So while I do think Abbrev is a funky little thing, I don't feel like there isn't way to get arguably more sane behavior. I'll have to experiment some more.
